qa: fix an off-by-one in utxo snapshot fuzz target and sanity check its snapshot data

darosior opened the pull request:
The assumeutxo data for the fuzz target could change and invalidate the hash silently, preventing the fuzz target from reaching some code paths. Fix this by introducing a unit test which would break if the snapshot data the fuzz target relies on were to change.

In implementing this i noticed the height used for coins in the fuzz target is actually off-by-one (as if the first block in the created chain was the genesis but it's block `1`), so fix that too.

darosior committed:
qa: correct off-by-one in utxo snapshot fuzz target

The chain starts at block 1, not genesis.

darosior commented:
cc @mzumsande. This was also suggested by @ryanofsky in https://github.com/bitcoin/bitcoin/pull/30329#discussion_r1671160670.

dergoegge reviewed (approved):
~~utACK a6d5ef3f604e646ae56c3a5c1aada48f9e134435~~

maflcko reviewed (commented):
left a question. Also, did you check the coverage is not decreased? Ref: https://maflcko.github.io/b-c-cov/fuzz.coverage/src/test/fuzz/utxo_snapshot.cpp.gcov.html

dergoegge commented on src/test/validation_chainstatemanager_tests.cpp:
```diff
@@ -836,4 +838,26 @@ BOOST_FIXTURE_TEST_CASE(chainstatemanager_args, BasicTestingSetup)
+
+    // Mine the same 200 blocks as the `utxo_snapshot` fuzz target.
+    for (auto& block: CreateBlockChain(snapshot_height, m_node.chainman->GetParams())) {
+        MineBlock(m_node, block);
```
Actually valid PoW will pass the reduced fuzz PoW check as well. But you're right I think the block hashes in the chain produced in the fuzz test and the unit test will be different.

We could move the assertion to the init function in the fuzz test, that should fail right now (i think).

darosior replied:
The fuzz mock `hash[31] & 0x80 == 0` is equivalent to the pow check against regtest min difficulty used in the unit test. If one passes the other necessarily does. This mean both will mine the same chain. You can check this as they both arrive to the same block hash at height 200, with the same utxo set hash.

darosior replied:
Ah right.

darosior replied:
> ```c++
>     hash = uint256{"7fffff0000000000000000000000000000000000000000000000000000000001"};
>     Assert(CheckProofOfWork(hash, nBits, consensus)); // fails in the unit test, passes in the fuzz test
> ```

I don't think this actually passes in the fuzz test. The hex representation is in reversed byte order. `hash[31]` here is `0x7f`. If it had its highest bit set (that is `hash[31] & 0x80 != 0` then it would be necessarily be higher than the regtest max target. 

All values with the highest bit set would be higher than the regtest maximum target (which is `7fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff`). These are `8000000000000000000000000000000000000000000000000000000000000000` to `ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff`. Therefore i think my original point stands.

darosior replied:
You are right, i missed how the genesis block's `nBits` does not correspond to the `powLimit` of `7fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff` but instead encodes a rounded down target of `7fffff0000000000000000000000000000000000000000000000000000000000`. This means that all values from `7fffff0000000000000000000000000000000000000000000000000000000001` to `7fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff` would have the highest bit unset but also fail in the unit test. If the `nBits` didn't round down the maximum target the two would match exactly.

Thanks for the patience.

darosior committed:
test util: split up ConnectBlock from MineBlock

darosior committed:
fuzz: sanity check hardcoded snapshot in utxo_snapshot target

The assumeutxo data for the fuzz target could change and invalidate the hash silently, preventing
the fuzz target from reaching some code paths.

Fix this by sanity checking the snapshot values during initialization.

darosior commented:
Just pushed an update to have the sanity check in the fuzz target initialization instead. Despite the previous version being technically correct it was so for [non-obvious reasons](https://github.com/bitcoin/bitcoin/pull/31910#discussion_r1963724337) and it makes sense to have the sanity check next to where the value will be used.

You can check it serves its intended purpose with the following diff for instance:
```diff
diff --git a/src/test/util/mining.cpp b/src/test/util/mining.cpp
index e006a2da629..443254ca503 100644
--- a/src/test/util/mining.cpp
+++ b/src/test/util/mining.cpp
@@ -43,6 +43,7 @@ std::vector<std::shared_ptr<CBlock>> CreateBlockChain(size_t total_height, const
         coinbase_tx.vout.resize(1);
         coinbase_tx.vout[0].scriptPubKey = P2WSH_OP_TRUE;
         coinbase_tx.vout[0].nValue = GetBlockSubsidy(height + 1, params.GetConsensus());
+        coinbase_tx.nLockTime = height;
         coinbase_tx.vin[0].scriptSig = CScript() << (height + 1) << OP_0;
         block.vtx = {MakeTransactionRef(std::move(coinbase_tx))};
 ```

Running the fuzz target with this diff would make the invalidation of the snapshot value in chainparams explicit:
```
test/fuzz/utxo_snapshot.cpp:65 sanity_check_snapshot: Assertion `stats.hashBlock == cp_au_data.blockhash' failed.
Aborted
```

Whereas if you apply the same diff on master the target will happily run and never be able to find a valid snapshot.

> Also, did you check the coverage is not decreased?

I did not generate a full coverage report but did check it was still possible for the target to find a valid snapshot and also check the `cov:` value reported by libfuzzer was the same on master and this branch (4200 without sanitizers).

mzumsande reviewed (commented):
> I did not generate a full coverage report but did check it was still possible for the target to find a valid snapshot and also check the cov: value reported by libfuzzer was the same on master and this branch (4200 without sanitizers).

More strictly, I'd expect that a fuzz seed that generated a valid snapshot before this change would also generate one with the change - the loop iteration that isn't done anymore doesn't make use of fuzzed_data_provider. So existing seeds shouldn't be invalidated.

mzumsande commented on src/test/util/mining.cpp:
```diff
@@ -92,6 +92,11 @@ COutPoint MineBlock(const NodeContext& node, std::shared_ptr<CBlock>& block)
+    return ConnectBlock(node, block);
+}
+
+COutPoint ConnectBlock(const NodeContext& node, const std::shared_ptr<CBlock>& block)
```
I'd prefer to call this wrapper around `ProcessNewBlock` some other name (e.g. `ProcessBlock()` like `net_processing` does) to avoid confusion with the existing `ConnectBlock` in validation. The added function doesn't necessarily attempt to connect the block, for example if a conflicting block arrived / was connected in between mining and calling `ProcessNewBlock` the real `ConnectBlock` may never be called.

darosior replied:
Done.

mzumsande reviewed (commented):
Code Review ACK 63b534f97e591d4e107fd5148909852eb2965d27

fjahr reviewed (commented):
tACK 63b534f97e591d4e107fd5148909852eb2965d27

Reviewed the code and played around with the fuzzer changes a bit.

darosior commented on src/test/util/mining.h:
```diff
@@ -32,6 +32,11 @@ COutPoint MineBlock(const node::NodeContext&,
 COutPoint MineBlock(const node::NodeContext&, std::shared_ptr<CBlock>& block);
 
+/**
+ * Returns the generated coin (or Null if the block was invalid).
```
I figured its name already implies what it's doing and expecting (an entirely valid block, not only valid PoW), and didn't want to just add a comment like "Process a block" (akin to `PrepareBlock` right below whose docstring states "Prepare a block"). I felt only the return value was not clear from the function declaration, hence why specifying it in the docstring.
